Fix `PopoverAnchor` not taking precedence over `PopoverDisclosure` when placing Popover

bengry opened the pull request:
Fixes #3729

This is my first PR to this repository, so although I read the [contributing file](https://github.com/ariakit/ariakit/blob/main/contributing.md), I might've goofed or missed something. Please let me know if there's anything I should've done differently.

bengry committed:
fix PopoverAnchor not taking precedence over PopoverDisclosure when Popover is placed

bengry committed:
Create thin-eyes-cross.md

diegohaz commented:
Thanks for tackling this, @bengry!

I think this implementation might introduce a breaking change, as some folks could expect the `anchorElement` state in the popover store to reference the disclosure element.

As an alternative, we could continue to use the popover anchor logic within the disclosure component, but introduce a kind of internal priority prop. Here, the `usePopoverDisclosure` hook would have a lower priority compared to a component that directly uses `usePopoverAnchor` (like `PopoverAnchor`). This could involve passing an internal prop or something similar. I'm still pondering over this.

bengry commented:
> Thanks for tackling this, @bengry!
> 
> I think this implementation might introduce a breaking change, as some folks could expect the `anchorElement` state in the popover store to reference the disclosure element.

Is this documented behavior? I think the code would be much harder to reason about.

> As an alternative, we could continue to use the popover anchor logic within the disclosure component, but introduce a kind of internal priority prop. Here, the `usePopoverDisclosure` hook would have a lower priority compared to a component that directly uses `usePopoverAnchor` (like `PopoverAnchor`). This could involve passing an internal prop or something similar. I'm still pondering over this.

I tried implementing this, but I'm not sure what's the proper way to pass in or save an internal state like this in Ariakit, especially because the library exposes all the interfaces and hooks as it's public API it makes things a bit tricky.
Maybe if you can point me in some direction I could take it further.

Anyway, IMO, especially because the library is still in pre-v1, and a migration guide is added to the release notes (basically, use [my workaround](https://github.com/ariakit/ariakit/issues/3729#issue-2259012011), but the other way around, elements-wise) is enough.

bengry commented:
> Anyway, IMO, especially because the library is still in pre-v1, and a migration guide is added to the release notes (basically, use [my workaround](https://github.com/ariakit/ariakit/issues/3729#issue-2259012011), but the other way around, elements-wise) is enough.

@diegohaz Hey, any news on this? Would love to get this fix in (although not urgent since it does have a workaround, as you've mentioned).

nicosammito commented:
Any updates on this topic?

diegohaz committed:
Merge remote-tracking branch 'origin/main' into wf-pr-3743

diegohaz committed:
Add disclosureElement to popover position effect deps

The fix changed the positioning anchor to `anchorElement ?? disclosureElement`,
so the layout effect now reads disclosureElement; add it to the dependency
array (replacing a pre-existing duplicate popoverElement entry) so positioning
re-runs and autoUpdate re-wires when the disclosure element changes.

diegohaz commented:
Closing in favor of https://github.com/ariakit/ariakit/pull/6804
